DEVELOPMENT.md

```diff
@@ -0,0 +1,309 @@
+Ensure that you have all three binaries available and running on your system.
+
+# Compilation Database
+
+For projects with non-trivial build systems, like PowerDNS, `clangd` requires a [compilation database](https://clang.llvm.org/docs/JSONCompilationDatabase.html).
+
+Since PowerDNS' autotools-based build system does not have native support for generating such a database, an external tool like [Bear (the Build EAR)](https://github.com/rizsotto/Bear) can be used.
+
+Once you have `bear` installed, configure a build of your choice (either the PowerDNS `auth`, `recursor` or `dnsdist`) using `clang` and `clang++`:
+
+```sh
+make distclean    # Ensure we rebuild all files so that bear can pick them up.
+CC=clang CXX=clang++ ./configure --with-modules=gsqlite3 --disable-lua-records
+```
+
+We can now build PowerDNS using `bear` and `make` which produces a compilation database file named `compile_commands.json`:
+
+```sh
+bear --append -- make -j 8
+```
+
+Once the compilation database is generated, you can now move onto setting up an LSP client in your editor or IDE.
+
+Note that the process of generating the compilation database file only needs to be run when a file is added to the project or when build flags change (e.g. dependencies are added).
+
+# Editors
+
+## Emacs
+
+This section explains how to set up [Emacs](https://www.gnu.org/software/emacs/) with [LSP Mode](https://emacs-lsp.github.io/) for C/C++ development using `clangd` which supports [on-the-fly checking](https://www.flycheck.org/en/latest/) and [auto-completion](https://company-mode.github.io/), among many other features.
+
+Instructions for an alternative, more minimal, setup using [Eglot](https://github.com/joaotavora/eglot) [are also available](#minimal-emacs).
+
+Code snippets below should be added to your Emacs init file (e.g. `~/.config/emacs/init`).
+
+We'll start by enabling Emacs package repositories and declaring which packages we would like to have installed:
+
+```elisp
+(with-eval-after-load 'package
+ (setq package-archives
+  '(("gnu"   . "https://elpa.gnu.org/packages/")
+    ("melpa" . "https://melpa.org/packages/")))
+ (push 'company      package-selected-packages)
+ (push 'flycheck     package-selected-packages)
+ (push 'lsp-mode     package-selected-packages)
+ (push 'lsp-ui       package-selected-packages)
+ (push 'lsp-treemacs package-selected-packages))
+```
+
+To avoid restarting Emacs, you can evaluate that previous s-expression by pointing at the last parenthesis and using `C-x C-e` or selecting the block and using `M-x eval-region`.
+
+Once done, run `M-x package-refresh-contents`, then `M-x package-install-selected-packages`.
+This should install some packages for you.
+
+Now let's set up our common programming mode, this enables the following features:
+
+* Highlighting the current line.
+* Displaying line numbers.
+* Auto-inserting indentation.
+* Auto-inserting closing parenthesis, brackets, etc...
```

Comment:
This can't be right as is.

```suggestion
* Auto-inserting closing parentheses, brackets, etc...
```

This would be fine, it's inserting in the mass sense plural `)` and plural `]`s.

Reply:
You could possibly do this instead:
```suggestion
* Auto-inserting closing parenthesis, bracket, etc...
```

Inserting a single `)` or a single `]`.